# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: uriloader/exthandler/ExternalHelperAppChild.cpp
# Commit: baffa8a7b961
# Full Hash: baffa8a7b9618e9a02e4bd808adb71dead99cce5
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-11-12 09:43:07
# Regressor Bug: 1589270
# File Overlap Count: 1
# Description:
#   Bug 1589270 - Part 3: Convert nsExternalHelperApp to use BrowsingContext instead of nsIInterfaceRequestor. r=bzbarsky
#   
#   This also converts MaybeCloseWindowHelper, and results in the window close operations being always run in the parent (even without DocumentChannel).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49528
# ==============================================================================

diff -r 68bc2ae9b89a -r baffa8a7b961 uriloader/exthandler/ExternalHelperAppChild.cpp
--- a/uriloader/exthandler/ExternalHelperAppChild.cpp	Fri Nov 08 03:01:03 2019 +0000
+++ b/uriloader/exthandler/ExternalHelperAppChild.cpp	Fri Nov 08 04:35:05 2019 +0000
@@ -12,6 +12,7 @@
 #include "nsIFTPChannel.h"
 #include "nsIRequest.h"
 #include "nsIResumableChannel.h"
+#include "nsIPropertyBag2.h"
 #include "nsNetUtil.h"
 
 namespace mozilla {
@@ -64,20 +65,9 @@
   nsresult rv = mHandler->OnStartRequest(request);
   NS_ENSURE_SUCCESS(rv, NS_ERROR_UNEXPECTED);
 
-  // Calling OnStartRequest could cause mHandler to close the window it was
-  // loaded for. In that case, the BrowserParent in the parent context might
-  // then point to the wrong window. Re-send the window context along with
-  // either DivertToParent or SendOnStartRequest just in case.
-  nsCOMPtr<nsPIDOMWindowOuter> window =
-      do_GetInterface(mHandler->GetDialogParent());
-  NS_ENSURE_TRUE(window, NS_ERROR_NOT_AVAILABLE);
-
-  BrowserChild* browserChild = mozilla::dom::BrowserChild::GetFrom(window);
-  NS_ENSURE_TRUE(browserChild, NS_ERROR_NOT_AVAILABLE);
-
   nsCOMPtr<nsIDivertableChannel> divertable = do_QueryInterface(request);
   if (divertable) {
-    return DivertToParent(divertable, request, browserChild);
+    return DivertToParent(divertable, request);
   }
 
   nsCString entityID;
@@ -85,7 +75,7 @@
   if (resumable) {
     resumable->GetEntityID(entityID);
   }
-  SendOnStartRequest(entityID, browserChild);
+  SendOnStartRequest(entityID);
   return NS_OK;
 }
 
@@ -102,8 +92,7 @@
 }
 
 nsresult ExternalHelperAppChild::DivertToParent(
-    nsIDivertableChannel* divertable, nsIRequest* request,
-    BrowserChild* browserChild) {
+    nsIDivertableChannel* divertable, nsIRequest* request) {
   // nsIDivertable must know about content conversions before being diverted.
   MOZ_ASSERT(mHandler);
   mHandler->MaybeApplyDecodingForExtension(request);
@@ -115,7 +104,7 @@
   }
   MOZ_ASSERT(diverter);
 
-  if (SendDivertToParentUsing(diverter, browserChild)) {
+  if (SendDivertToParentUsing(diverter)) {
     mHandler->DidDivertRequest(request);
     mHandler = nullptr;
     return NS_OK;